lib/pure/strutils.nim

```diff
@@ -2050,97 +2050,152 @@ func contains*(s: string, chars: set[char]): bool =
+    inc i
+  if occLeft == 0:
+    # copy the rest:
+    result.add s.substr(i)
+
+# Maybe implement something like this for all replace functions...?
+# So that they can for example replace all Whitespace characters
+#[
+func replace*(s: string, subs: set[char], by: char, max = -1): string
+    {.rtl, extern: "nsuReplaceSetChar".} =
+  ## Replaces `max` occurrences of any characters in `subs` found
+  ## in `s` with the character `by`.
+  ## If `max` is set to `-1`, it replaces all occurrences.
+  ##
+  ## Optimized version of `replace <#replace,string,char[string],string>`_ for
+  ## characters.
+  ##
+  ## See also:
+  ## * `find func<#find,string,char,Natural,int>`_
+  ## * `replaceWord func<#replaceWord,string,string,string>`_
+  ## * `multiReplace func<#multiReplace,string,varargs[]>`_
+  result = newString(s.len)
+  var occLeft = max
+  for i in 0..s.high:
+    if occLeft != 0 and s[i] in subs:
+      result[i] = by
+      if occLeft > 0: occLeft -= 1
+    else:
+      result[i] = s[i]
+]#
+
+func replace*(s, sub: string, by = "", max = -1): string {.rtl,
     extern: "nsuReplaceStr".} =
-  ## Replaces every occurrence of the string `sub` in `s` with the string `by`.
+  ## Replaces `max` occurrences of the string `sub` in `s` with the
+  ## string `by`.
+  ## If `max` is set to `-1`, it replaces all occurrences.
   ##
   ## See also:
   ## * `find func<#find,string,string,Natural,int>`_
   ## * `replace func<#replace,string,char,char>`_ for replacing
   ##   single characters
   ## * `replaceWord func<#replaceWord,string,string,string>`_
   ## * `multiReplace func<#multiReplace,string,varargs[]>`_
+  runnableExamples:
+    assert "FooFooBaz".replace("Foo", "Bar") == "BarBarBaz"
+    assert "BarBarBaz".replace("Bar", "Foo", 1) == "FooBarBaz"
   result = ""
   let subLen = sub.len
-  if subLen == 0:
-    result = s
+  if subLen == 0: result = s
   elif subLen == 1:
+    let subChar = sub[0]
     # when the pattern is a single char, we use a faster
     # char-based search that doesn't need a skip table:
-    let c = sub[0]
-    let last = s.high
-    var i = 0
+    var
+      i = 0
+      occLeft = max
     while true:
-      let j = find(s, c, i, last)
-      if j < 0: break
-      add result, substr(s, i, j - 1)
-      add result, by
+      let j = s.find(subChar, i)
+      if j == -1 or occLeft == 0: break
+      if occLeft > 0: dec occLeft
+
+      result.add s.substr(i, j - 1)
+      result.add by
       i = j + subLen
     # copy the rest:
-    add result, substr(s, i)
+    result.add s.substr(i)
   else:
     var a {.noinit.}: SkipTable
     initSkipTable(a, sub)
-    let last = s.high
-    var i = 0
+    var
+      i = 0
+      occLeft = max
     while true:
-      let j = find(a, s, sub, i, last)
-      if j < 0: break
-      add result, substr(s, i, j - 1)
-      add result, by
+      let j = find(a, s, sub, i)
+      if j == -1 or occLeft == 0: break
+      if occLeft > 0: dec occLeft
+      result.add s.substr(i, j - 1)
```

Comment:
pre-existing but this creates un-necessary allocations.

IMO we need a 0-allocation `addReplace` in strbasics and then strutils.replace can reuse it.

not a blocker for this PR though as this refactoring can be done later.

Reply:
We should use a TR macro to optimize this, the pattern is common and while we can avoid it, many of our users don't know about it.

Reply:
that won't make `replace` 0-alloc though, hence the need for `strbasics.addReplace` in future work.

And for that particular line, all that's needed is replace `substr` by `toOpenArray`:
```
      result.add s.substr(i, j - 1)
=>
      result.add s.toOpenArray(i, j - 1)
```
(with `import std/strbasics` on top), which is now possible since #15951